File: src/main/java/services/java/ItemService.java
package services.java;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class ItemService {

    private final Random random = new Random();
    
    /** random prices */
    private final double[] prices = {
      10.0,
      20.0,
      30.0,
      40.0,
      50.0,
      60.0,
      70.0,
      80.0,
      90.0
    };
    
    public List<Item> getItems(int client) {
        List<Item> items = new ArrayList<>();
        
        for(int i = 1; i <= 50; i++) {
            String id = Long.toHexString(random.nextLong());
            double price = prices[random.nextInt(9)];
            
            Item item = new Item(id, client, price);
            items.add(item);
            
            // Throw in some randoming waiting.
            // DB, webservices or shitty hardware
            try {
                Thread.sleep((long) (random.nextDouble() * 100L));
            } catch (InterruptedException e) {
                throw new RuntimeException("Why wake up?", e);
            }
        }
        System.out.println(Thread.currentThread().getName() + " : finished");
        return items;
    }
}

File: src/main/scala/services/scala/ItemService.scala
package services.scala

import scala.language.{ postfixOps }
import scala.util.Random.{ nextLong, nextInt, nextDouble }
import scala.collection.immutable.Seq

case class Item(id: String, client: Int, price: Double)

class ItemService {

  val prices = (10.0 to 100.0 by 10)

  def getItems(client: Int): Seq[Item] = {
    val result = (0 to 50).map { _ =>
      Item(id = nextLong.toHexString, client = client, price = prices(math.abs(nextInt) % 9))
    }.toSeq
    // waiting for db, webservices,...
    Thread sleep (nextDouble * 100.0).toLong
    println(s"${Thread.currentThread.getName} : finished")
    result
  }

}

class FlakyItemService extends ItemService {
  
  /**
   * creates random errors plus a stable error for clientId == 2
   */
  override def getItems(client: Int) : Seq[Item] = {
    val result = super.getItems(client)
    
    // flakiness
    if(nextDouble < 0.2 || client == 2) {
      throw new IllegalStateException(s"Something went wrong for client $client")
    }
    
    result
  }
  
}